Feat/arc ingest command handler support

cain4chain opened the pull request:
## Description of Changes
feat(arc_ingest.go) : add main arc-ingest method
feat(jsonutil.go) : add new JSON util method to parse byte array and handle errors
feat(limited_body_reader.go) : Add specified utility method to read data chunks, will be used in other command handlers such as submit_transaction.go

## Linked Issues / Tickets
Closes : Add arc-ingest command handler support #17

## Testing Procedure
feat(arc_ingest_test.go) : Add happy test and table driven tests for validation errors
feat(limited_body_reader_test) : Test cases for testing limited_body_reader.go method.

- [x] I have added new unit tests
- [x] All tests pass locally
- [x] I have tested manually in my local environment

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have run the linter

cain4chain committed:
feat(arc_ingest_command_handler) : Integrate arc-ingest command handler code

feat(contract.go) : Implement HandleNewMerkleProof method into OverlayEngineProvider
feat(http.go) : Add arc-ingest route to http file
feat(app.go) : Configure ArcIngestHandler into app commands
feat(arc_ingest.go) : add main arc-ingest method
feat(arc_ingest_test.go) : Add happy test and table driven tests for validation errors
feat(jsonutil.go) : add new JSON util method to parse byte array and handle errors
feat(limited_body_reader.go) : Add specified utility method to read data chunks, will be used in other command handlers such as submit_transaction.go
feat(limited_body_reader_test) : Test cases for testing limited_body_reader.go method.
chore(NoopEngineProvider) : Add HandleNewMerkleProof method to noop engine provider

github-actions[bot] committed:
chore: update generated files

cain4chain committed:
chore(lint) : line spacing, add custom logging

chore(lint) : fix line spacing issue
chore(slog.Error) : update custom logging value for merkle proof processing

github-actions[bot] committed:
chore: update generated files

cain4chain committed:
feat(arc_ingest_handler) : tidy up comments

chore(arc_ingest) : Remove unrequired comments

github-actions[bot] committed:
chore: update generated files

cain4chain committed:
Merge branch 'main' into feat/arc-ingest-command-handler-support

cain4chain committed:
Merge branch 'main' into feat/arc-ingest-command-handler-support

chris-4chain commented on pkg/server/internal/app/commands/arc_ingest.go:
```diff
@@ -0,0 +1,269 @@
+	}
+	var hash chainhash.Hash
+	copy(hash[:], bb)
+	return hash, nil
```
You could use a function from `go-sdk`: [NewHashFromHex](https://github.com/bsv-blockchain/go-sdk/blob/dbbb2cde494d1852eb6e312b286aa9ef975bbbf0/chainhash/hash.go#L110)
- It does necessary validation also.

chris-4chain commented on pkg/server/internal/app/commands/arc_ingest.go:
```diff
@@ -0,0 +1,269 @@
+		default:
+			jsonutil.SendHTTPResponse(w, http.StatusBadRequest, NewFailureArcIngestHandlerResponse(err.Error()))
+			return
+		}
```
This could be moved to private helper function.

chris-4chain commented on pkg/server/internal/app/commands/arc_ingest.go:
```diff
@@ -0,0 +1,269 @@
+// Handle processes an ARC ingest request, handling the validation, converting
+// the request data into the correct format, and invoking the provider to handle
+// the new Merkle proof. It also manages timeout and error responses.
+func (h *ArcIngestHandler) Handle(w http.ResponseWriter, r *http.Request) {
```
Maybe, such methods should receive a `ctx context.Context` as a first argument. 
- And correctly react when provided context is "canceled" (because of timeout, for example)

mgosek-4chain replied:
It would be better to create a decorator instead of duplicating code across each handler's codebase. However, we can add that "catch" in a separate PR. Sounds good? @chris-4chain

chris-4chain commented on pkg/server/internal/app/commands/arc_ingest.go:
```diff
@@ -0,0 +1,269 @@
+		if err == nil {
+			jsonutil.SendHTTPResponse(w, http.StatusOK, NewSuccessArcIngestHandlerResponse())
+			return
+		}
```
I suppose, it could be made by [context.WithTimeout ](https://golang.cafe/blog/golang-context-with-timeout-example.html).

cain4chain replied:
Ok thanks I will update this.

mgosek-4chain replied:
@cain4chain 

```
	ctx, cancel := context.WithTimeout(r.Context(), h.responseTimeout)
	defer cancel()

	err = h.provider.HandleNewMerkleProof(ctx, &txIDHash, merklePath)
	switch {
	case errors.Is(err, context.DeadlineExceeded):
		jsonutil.SendHTTPResponse(w, http.StatusGatewayTimeout, NewFailureArcIngestHandlerResponse(http.StatusText(http.StatusGatewayTimeout)))

	case errors.Is(err, context.Canceled):
		jsonutil.SendHTTPResponse(w, http.StatusRequestTimeout, NewFailureArcIngestHandlerResponse(http.StatusText(http.StatusRequestTimeout)))

	case err != nil:
		jsonutil.SendHTTPResponse(w, http.StatusInternalServerError, NewFailureArcIngestHandlerResponse(http.StatusText(http.StatusInternalServerError)))

	default:
		jsonutil.SendHTTPResponse(w, http.StatusOK, NewSuccessArcIngestHandlerResponse())
	}
```

Additionally, it would be good to add some structured logging and extract it into a separate function.

chris-4chain commented on pkg/server/internal/app/commands/arc_ingest.go:
```diff
@@ -0,0 +1,269 @@
+	h := ArcIngestHandler{
+		provider:         provider,
+		requestBodyLimit: jsonutil.RequestBodyLimit1GB,
+		responseTimeout:  10 * time.Second,
```
Please, make a constant for it.

chris-4chain commented on pkg/server/internal/app/jsonutil/limited_body_reader.go:
```diff
@@ -0,0 +1,64 @@
+func (l *LimitedBodyReader) Read() ([]byte, error) {
+	reader := io.LimitReader(l.Body, l.ReadLimit+1)
+	buff := bytes.NewBuffer(nil)
+	bb := make([]byte, 64*1024)
```
It looks like "magic number" - please make a constant with descriptive name for it. 
- e.g. `chunkSize`

mgosek-4chain commented on pkg/server/internal/app/commands/arc_ingest_test.go:
```diff
@@ -0,0 +1,193 @@
+	"github.com/stretchr/testify/require"
+)
+
+// Test_ArcIngestHandler_ShouldRespondsWith200AndCallsProvider tests the successful handling of a valid request
```
According to our convention, we don't add comments to unit tests, so please remove them :)

mgosek-4chain commented on pkg/server/internal/app/commands/arc_ingest.go:
```diff
@@ -0,0 +1,269 @@
+	// ErrMissingRequiredMerklePathFieldDefinition is returned when the request body is missing
+	// the required Merkle path field.
+	ErrMissingRequiredMerklePathFieldDefinition = errors.New("missing required field: merkle path")
+)
```
```
var (
	// ErrInvalidTxIDFormat is returned when the transaction ID is not in a valid format (e.g., not hexadecimal).
	ErrInvalidTxIDFormat = errors.New("invalid transaction ID format")

	// ErrInvalidTxIDLength is returned when the transaction ID does not match the expected length (typically 64 characters for a SHA-256 hash).
	ErrInvalidTxIDLength = errors.New("invalid transaction ID length")

	// ErrInvalidMerklePathFormat is returned when the Merkle path is malformed or does not conform to the expected structure.
	ErrInvalidMerklePathFormat = errors.New("invalid Merkle path format")

	// ErrMissingRequiredRequestFieldsDefinition is returned when the request body is missing
	// required fields, such as the transaction ID and Merkle path.
	ErrMissingRequiredRequestFieldsDefinition = errors.New("missing required fields: txid, merkle path")

	// ErrMissingRequiredTxIDFieldDefinition is returned when the request body is missing
	// the required transaction ID field.

	ErrMissingRequiredTxIDFieldDefinition = errors.New("missing required field: txid")

	// ErrMissingRequiredMerklePathFieldDefinition is returned when the request body is missing
	// the required Merkle path field.
	ErrMissingRequiredMerklePathFieldDefinition = errors.New("missing required field: merkle path")
)
```

Added missing spaces.

mgosek-4chain commented on pkg/server/internal/app/commands/arc_ingest_test.go:
```diff
@@ -0,0 +1,193 @@
+			require.Contains(t, response.Message, tc.expectedError.Error())
+			require.Equal(t, "error", response.Status)
+		})
+	}
```
```
	tests := map[string]struct {
		method             string
		payload            commands.ArcIngestRequest
		expectedResponse   commands.ArcIngestHandlerResponse
		expectedHTTPStatus int
	}{
		"should fail with 405 when HTTP method is GET": {
			method: http.MethodGet,
			payload: commands.ArcIngestRequest{
				TxID:        testutil.ValidTxId,
				MerklePath:  testutil.NewValidTestMerklePath(t),
				BlockHeight: 848372,
			},
			expectedResponse:   commands.NewFailureArcIngestHandlerResponse(commands.ErrInvalidHTTPMethod.Error()),
			expectedHTTPStatus: http.StatusMethodNotAllowed,
		},

		"should fail with 400 when all required fields are missing": {
			method: http.MethodPost,
			payload: commands.ArcIngestRequest{
				TxID:        "",
				MerklePath:  "",
				BlockHeight: 0,
			},
			expectedResponse:   commands.NewFailureArcIngestHandlerResponse(commands.ErrMissingRequiredRequestFieldsDefinition.Error()),
			expectedHTTPStatus: http.StatusBadRequest,
		},
		"should fail with 400 when TxID field is missing": {
			method: http.MethodPost,
			payload: commands.ArcIngestRequest{
				TxID:        "",
				MerklePath:  testutil.NewValidTestMerklePath(t),
				BlockHeight: 848372,
			},
			expectedResponse:   commands.NewFailureArcIngestHandlerResponse(commands.ErrMissingRequiredTxIDFieldDefinition.Error()),
			expectedHTTPStatus: http.StatusBadRequest,
		},
		"should fail with 400 when MerklePath field is missing": {
			method: http.MethodPost,
			payload: commands.ArcIngestRequest{
				TxID:        testutil.ValidTxId,
				MerklePath:  "",
				BlockHeight: 848372,
			},
			expectedResponse:   commands.NewFailureArcIngestHandlerResponse(commands.ErrMissingRequiredMerklePathFieldDefinition.Error()),
			expectedHTTPStatus: http.StatusBadRequest,
		},
		"should fail with 400 when TxID format is invalid": {
			method: http.MethodPost,
			payload: commands.ArcIngestRequest{
				TxID:        "invalid-hex-string",
				MerklePath:  testutil.NewValidTestMerklePath(t),
				BlockHeight: 848372,
			},
			expectedResponse:   commands.NewFailureArcIngestHandlerResponse(commands.ErrInvalidTxIDFormat.Error()),
			expectedHTTPStatus: http.StatusBadRequest,
		},
		"should fail with 400 when TxID length is invalid": {
			method: http.MethodPost,
			payload: commands.ArcIngestRequest{
				TxID:        "1234",
				MerklePath:  testutil.NewValidTestMerklePath(t),
				BlockHeight: 848372,
			},
			expectedResponse:   commands.NewFailureArcIngestHandlerResponse(commands.ErrInvalidTxIDLength.Error()),
			expectedHTTPStatus: http.StatusBadRequest,
		},
		"should fail with 400 when MerklePath format is invalid": {
			method: http.MethodPost,
			payload: commands.ArcIngestRequest{
				TxID:        testutil.ValidTxId,
				MerklePath:  "invalid-merkle-path",
				BlockHeight: 848372,
			},
			expectedResponse:   commands.NewFailureArcIngestHandlerResponse(commands.ErrInvalidMerklePathFormat.Error()),
			expectedHTTPStatus: http.StatusBadRequest,
		},
	}

	for name, tc := range tests {
		t.Run(name, func(t *testing.T) {
			// given:
			mock := testutil.NewMerkleProofProviderMock(nil, tc.payload.BlockHeight)
			handler, err := commands.NewArcIngestHandler(mock)
			require.NoError(t, err)

			ts := httptest.NewServer(http.HandlerFunc(handler.Handle))
			defer ts.Close()

			req, err := http.NewRequest(tc.method, ts.URL, testutil.RequestBody(t, tc.payload))
			require.NoError(t, err)

			// when:
			res, err := ts.Client().Do(req)
			require.NoError(t, err)
			defer res.Body.Close()

			// then:
			require.Equal(t, tc.expectedHTTPStatus, res.StatusCode)

			var actualResponse commands.ArcIngestHandlerResponse
			err = jsonutil.DecodeResponseBody(res, &actualResponse)
			require.NoError(t, err)

			require.Equal(t, tc.expectedResponse, actualResponse)
		})
	}
```

mgosek-4chain commented on pkg/server/internal/app/commands/arc_ingest_test.go:
```diff
@@ -0,0 +1,193 @@
+			},
+			expectedStatus: http.StatusMethodNotAllowed,
+			expectedError:  commands.ErrInvalidHTTPMethod,
+		},
```
One of them is sufficient; we don't need both testing the same edge case :)

cain4chain committed:
chore : update as per review

chore(limit_body_reader) : set const for data chunk size
chore(arc_ingest_test) : optimize test cases, Add response struct validation
chore(arc_ingest) : optimize txID error validation, consolidate switch statements in individual methods, set const for default response time

github-actions[bot] committed:
chore: update generated files

cain4chain committed:
chore : update error messages

chore(arc_ingest) : update error messages for handling merkle proof results

github-actions[bot] committed:
chore: update generated files

cain4chain committed:
Merge branch 'main' into feat/arc-ingest-command-handler-support

mgosek-4chain commented on pkg/server/internal/app/commands/arc_ingest.go:
```diff
@@ -0,0 +1,306 @@
+// sending appropriate HTTP responses based on the specific error type.
+func (h *ArcIngestHandler) handleMerkleProofResult(w http.ResponseWriter, err error) {
+	switch {
+	case errors.Is(err, context.DeadlineExceeded):
+		slog.Error(fmt.Sprintf("[ArcIngest] Merkle proof processing timed out: %v", err))
+		jsonutil.SendHTTPResponse(w, http.StatusGatewayTimeout, NewFailureArcIngestHandlerResponse(ErrMerkleProofProcessingTimeout.Error()))
+
+	case errors.Is(err, context.Canceled):
+		slog.Error(fmt.Sprintf("[ArcIngest] Merkle proof processing canceled: %v", err))
+		jsonutil.SendHTTPResponse(w, http.StatusRequestTimeout, NewFailureArcIngestHandlerResponse(ErrMerkleProofProcessingCanceled.Error()))
```
Minor: could you please add unit tests to cover both timeout and cancellation scenarios?

chris-4chain commented on pkg/server/internal/app/commands/arc_ingest.go:
```diff
@@ -0,0 +1,306 @@
+
+// decode reads and decodes the request body into the provided destination
+// struct. It ensures the request body size is within the allowed limit.
+func (h *ArcIngestHandler) decode(body io.Reader, dst *ArcIngestRequest) error {
```
FYI:
- Not exported methods don't need doc-comment - add it only if it adds valuable context 
- Things in `internal` don't need doc-comment

Here we have both - not exported (private) and in internal 🙈

cain4chain replied:
Ok thanks will update.

cain4chain committed:
chore: add missing test cases

chore(arc_ingest_test) : add missing test cases for timeout scenarios
